src/server/pfs/server/driver.go

```diff
@@ -793,30 +789,47 @@ func (d *driver) getCompactedDiffFileSet(ctx context.Context, commit *pfs.Commit
 // The ProjectInfo provided to the closure is repurposed on each invocation, so it's the client's responsibility to clone the ProjectInfo if desired
 func (d *driver) listProject(ctx context.Context, cb func(*pfs.ProjectInfo) error) error {
 	authIsActive := true
-	projectInfo := &pfs.ProjectInfo{}
-	return errors.Wrap(d.projects.ReadOnly(ctx).List(projectInfo, col.DefaultOptions(), func(string) error {
+	projIter, err := coredb.ListProject(ctx, d.env.DB)
+	if err != nil {
+		return errors.Wrap(err, "could not list projects")
+	}
+	projInfo := &pfs.ProjectInfo{}
+	for !errors.Is(projIter.Next(ctx, projInfo), io.EOF) {
 		if authIsActive {
-			resp, err := d.env.AuthServer.GetPermissions(ctx, &auth.GetPermissionsRequest{Resource: projectInfo.GetProject().AuthResource()})
+			resp, err := d.env.AuthServer.GetPermissions(ctx, &auth.GetPermissionsRequest{Resource: projInfo.GetProject().AuthResource()})
 			if err != nil {
 				if errors.Is(err, auth.ErrNotActivated) {
 					// Avoid unnecessary subsequent Auth API calls.
 					authIsActive = false
-					return cb(projectInfo)
+					if err := cb(projInfo); err != nil {
+						return errors.Wrapf(err, "error getting permissions for project %s", projInfo.Project)
+					}
+					continue
 				}
-				return errors.Wrapf(err, "error getting permissions for project %s", projectInfo.Project)
+				return errors.Wrapf(err, "error getting permissions for project %s", projInfo.Project)
 			}
-			projectInfo.AuthInfo = &pfs.AuthInfo{Permissions: resp.Permissions, Roles: resp.Roles}
+			projInfo.AuthInfo = &pfs.AuthInfo{Permissions: resp.Permissions, Roles: resp.Roles}
 		}
-		return cb(projectInfo)
-	}), "could not list projects")
+		if err := cb(projInfo); err != nil {
+			return errors.Wrapf(err, "could not execute callback on project %s", projInfo.Project.Name)
+		}
+	}
+	return nil
 }
 
 // The ProjectInfo provided to the closure is repurposed on each invocation, so it's the client's responsibility to clone the ProjectInfo if desired
 func (d *driver) listProjectInTransaction(txnCtx *txncontext.TransactionContext, cb func(*pfs.ProjectInfo) error) error {
 	projectInfo := &pfs.ProjectInfo{}
-	return errors.EnsureStack(d.projects.ReadWrite(txnCtx.SqlTx).List(projectInfo, col.DefaultOptions(), func(string) error {
-		return cb(projectInfo)
-	}))
+	projIter, err := coredb.ListProject(txnCtx.Context(), d.env.DB)
+	if err != nil {
+		return errors.Wrap(err, "could not list project")
+	}
+	for !errors.Is(projIter.Next(txnCtx.Context(), projectInfo), io.EOF) {
```

Comment:
You can probably replace this with `stream.ForEach`.  I think the callback type would match exactly?

Reply:
I don't think this handles errors correctly, either.  Non-EOS errors would cause the loop to execute and call cb with bogus.  Switching to ForEach would handle those cases correctly.